block/src/Pos/Block/Network/Retrieval.hs

```diff
@@ -308,3 +309,53 @@ getProcessBlocks pm diffusion nodeId desired checkpoints = do
                   else pure False
           when exitedRecovery $
               logInfo "Recovery mode exited gracefully on receiving block we needed"
+
+-- Attempts to catch up by streaming blocks from peer.
+-- Will fall back to getProcessBlocks if streaming is disabled
+-- or not supported by peer.
+streamProcessBlocks
+    :: forall ctx m.
+       ( BlockWorkMode ctx m
+       , HasMisbehaviorMetrics ctx
+       )
+    => ProtocolMagic
+    -> Diffusion m
+    -> NodeId
+    -> HeaderHash
+    -> [HeaderHash]
+    -> m ()
+streamProcessBlocks pm diffusion nodeId desired checkpoints = do
+    logInfo "streaming start"
+    r <- Diffusion.streamBlocks diffusion nodeId desired checkpoints (loop (0::Word32) [])
+    case r of
+         Nothing -> do
+             logInfo "streaming not supported, reverting to batch mode"
+             getProcessBlocks pm diffusion nodeId desired checkpoints
+         Just _  -> do
+             logInfo "streaming done"
+             return ()
+  where
+    loop !n !blocks (streamWindow, wqgM, blockChan) = do
+        streamEntry <- atomically $ readTBQueue blockChan
+        case streamEntry of
+          StreamEnd         -> addBlocks blocks
+          StreamBlock !block -> do
```

Comment:
Maybe the bang here should be moved inside `StreamEntry` type declaration, that would be just more consistent with the rest of the code base.  I think we had a policy for that, I am right @avieth?

Reply:
Will move it to `StreamEntry`.